### DISCUSSION TITLE
[Question]Precise control of generating a geological model

### DISCUSSION THREAD

**Discussion by Changwan-planet** [gempy-project/gempy#927] [Q&A]
@javoha 

I Really appreciate opening your program and giving us the nice, quick feedback from this discussion.

I like my model created from gempy, specially it can make nice unconformity deformed by the reverse fault. 
Please refer to the uploaded *.zip file including Points, Orientations, and Jupyter notebook.
However, because the unconformity has some eroded surface, I would like to correct some specific points in the model.
![image](https://github.com/user-attachments/assets/7eee5ad3-81aa-422d-b889-d120b719fd95)

There are three issues. 

=========================================================================================

1. As you can see the image below, **the model is curved at the outside of the data points** (the blue ellipses).  
   That is not my intention. How to make it look good or to follow the trend of the model?
<img width="1221" alt="image" src="https://github.com/user-attachments/assets/51242453-314f-4e8e-8761-fe6c7b00f3a1">

=========================================================================================
2. **The generated surface and the points are not contact with each other**.  
    That part is the eroded surface (the red circle).

<img width="1278" alt="image" src="https://github.com/user-attachments/assets/1768d84f-3b29-466b-a1cc-cbae86a8ab60">

<img width="1280" alt="image" src="https://github.com/user-attachments/assets/6db3769a-bec7-4745-a2d9-e46c028b73aa">

     
I know if i change this value 

geo_data.interpolation_options.kernel_options.range,

I could get the goal.
But if that value is too small like below 0.3, the model is too distorted and ugly. 

I also know every point is not easy to contact the modeled surface with my best satisfaction.
**If I want to make the certain part of the surface contact with the point, Which action is better?** 

=========================================================================================
3. From time to time, **my model has some holes** (the red circle) when I change the above value as 0.1 . 
    Why do I get the hole and how to remove the hole with the above value constant?  
    Even though I increased the resolution, there was no change.

![image](https://github.com/user-attachments/assets/a6226b29-087e-4db8-b795-fcd5aad4b914)

[Miho_Model.zip](https://github.com/user-attachments/files/16443032/Miho_Model.zip)

Thank you so much!!

**Comment by javoha** [comment]
Hi @Changwan-planet,
thanks for your questions. I am out of office at the moment so I can't take a close look but I can already give some general answers that hopefully help:

1. The underlying co-kriging algorithm tends to curve, especially in areas without data. You could either enter additional control points by hand to manually stop this behaviour or alternatively use some kind of transformation to model an anisotropic (directional) variogram. I would suggest using the first option for a quick fix, as the second involves more work and might mess with your model.

2. and 3. kind of belong together: surfaces don't match exactly to points due to resolution. It is important to remember that the main result of a model is the block model. This is what you define your resolution for. Depending on that resolution each point of the resulting regular grid gets a ID that corresponds to a structural element. 
The surfaces are extracted from this lithology block model mainly for visualisation. In gempy v3 dual contouring is used for this process and it is not perfect yet. Surfaces not going exactly through points is a result of this and also the holes are a result of the approach that we are trying to fix at the moment. We also plan to bring back marching cubes as an alternative algorithm.
It is important to note that this is not because there is a problem with your model, it is really just an issue with the surface extracting approach.

Hope this helps and maybe I can take a closer look in a couple of weeks.

Cheers 
Jan

**Reply by Changwan-planet** [reply]
Hi, @javoha, 
Thanks for your reply in detail even though you are out of your office!!
I will try something that I can do based on your answer!
